beacon-chain/cache/depositcache/deposits_cache.go

```diff
@@ -91,9 +104,65 @@ func (dc *DepositCache) InsertDepositContainers(ctx context.Context, ctrs []*dbp
 	historicalDepositsCount.Add(float64(len(ctrs)))
 }
 
-// AllDepositContainers returns a list of deposits all historical deposit containers until the given block number.
+// InsertFinalizedDeposits inserts deposits up to eth1DepositIndex (exclusive) into the finalized deposits cache.
+func (dc *DepositCache) InsertFinalizedDeposits(ctx context.Context, eth1DepositIndex int64) {
+	ctx, span := trace.StartSpan(ctx, "DepositsCache.InsertFinalizedDeposits")
+	defer span.End()
+	dc.depositsLock.Lock()
+	defer dc.depositsLock.Unlock()
+
+	var depositTrie *trieutil.SparseMerkleTrie
+
+	if dc.finalizedDeposits != nil {
+		depositTrie = dc.finalizedDeposits.Deposits
+
+		insertIndex := dc.finalizedDeposits.MerkeTreeIndex + 1
+		for _, d := range dc.deposits {
+			if d.Index <= dc.finalizedDeposits.MerkeTreeIndex {
+				continue
+			}
+			if d.Index >= eth1DepositIndex {
+				break
+			}
+			depHash, err := ssz.HashTreeRoot(d.Deposit.Data)
+			if err != nil {
+				log.WithError(err).Error("Could not hash deposit data. Finalized deposit cache not updated.")
+				return
+			}
+			depositTrie.Insert(depHash[:], int(insertIndex))
+			insertIndex++
+		}
+	} else {
+		var finalizedDeposits [][]byte
+		for _, d := range dc.deposits {
+			if d.Index >= eth1DepositIndex {
+				break
+			}
+			hash, err := ssz.HashTreeRoot(d.Deposit.Data)
+			if err != nil {
+				log.WithError(err).Error("Could not hash deposit data. Finalized deposit cache not updated.")
+				return
+			}
+			finalizedDeposits = append(finalizedDeposits, hash[:])
+		}
+
+		var err error
+		depositTrie, err = trieutil.GenerateTrieFromItems(finalizedDeposits, int(params.BeaconConfig().DepositContractTreeDepth))
+		if err != nil {
+			log.WithError(err).Error("Could not generate deposit trie. Finalized deposit cache not updated.")
+			return
+		}
+	}
+
+	dc.finalizedDeposits = &FinalizedDeposits{
+		Deposits:       depositTrie,
+		MerkeTreeIndex: eth1DepositIndex - 1,
```

Comment:
Why -1? Shouldn't it match?

Reply:
I don't know 😕 

Please take a look at my first comment. The two snippets with `>=` comparison treat the deposit whose index is equal to state's `Eth1DepositIndex` as a pending deposit. I am not sure why... I implemented the cache in such a way that the deposit with such an index is NOT a finalized deposit. There are other places where this needs to be changed if we agree that this is incorrect.

It would be great if someone could help out here and explain if the logic in the snippets is correct or not.